app/main/views/process_outcomes.py

```diff
@@ -0,0 +1,86 @@
+import datetime
+
+from flask import abort
+from dmapiclient.audit import AuditTypes
+
+from .. import main
+from ... import db
+from ...models import AuditEvent, ProcessOutcome
+
+from ...utils import (
+    get_json_from_request,
+    json_has_required_keys,
+    single_result_response,
+    validate_and_return_updater_request,
+)
+from ...validation import validate_process_outcome_json_or_400
+
+
+@main.route("/process-outcomes/<int:process_outcome_id>", methods=("GET",))
+def get_process_outcome(process_outcome_id):
+    process_outcome = ProcessOutcome.query.filter_by(external_id=process_outcome_id).first_or_404()
+
+    return single_result_response("processOutcome", process_outcome), 200
+
+
+@main.route("/process-outcomes/<int:process_outcome_id>", methods=("PUT",))
+def update_process_outcome(process_outcome_id):
+    uniform_now = datetime.datetime.utcnow()
+
+    updater_json = validate_and_return_updater_request()
+
+    json_payload = get_json_from_request()
+    json_has_required_keys(json_payload, ["processOutcome"])
+    process_outcome_json = json_payload["processOutcome"]
+
+    validate_process_outcome_json_or_400(process_outcome_json)
+
+    # fetch and lock ProcessOutcome row so we know writing this back won't overwrite any other updates to it that made
+    # it in in the meantime
+    process_outcome = db.session.query(ProcessOutcome).filter_by(
+        external_id=process_outcome_id,
+    ).with_for_update().first()
+
+    if not process_outcome:
+        abort(404, f"Process outcome {process_outcome_id} not found")
+
+    process_outcome.update_from_json(process_outcome_json)
+
+    if process_outcome.completed_at is not None and process_outcome_json.get("completed") is False:
+        abort(400, f"Can't un-complete process outcome")
+    if process_outcome.completed_at is None and process_outcome_json.get("completed") is True:
+        # make a cursory check for existing ProcessOutcome collisions. we're not actually totally relying on this
```

Comment:
This comment is lengthy. You could just say we're checking to determine which error message to show 🤷‍♀️

Reply:
I think it's fair enough to note that this is technically duplicating something that exists elsewhere, and why.

Reply:
Lengthy comments are fine if they're relevant and in fact something that I think are sorely lacking in our code making it very hard to have confidence about the intention of the design.

Reply:
But this comment, specifically, is not concise. It uses too many words to say not enough important stuff.

From a 'why I wrote my code like this' point of view it's good. You do the check here instead of relying on the db constraint to return a more readable error message. Because otherwise someone might question why you're doing the check instead of relying on the constraint. Cool.

With your code comments it's worth putting yourself in the shoes of 'someone reading the code just to find out what it does' point of view. When I read this code I can see that you're doing a constraint check and raising an error. That's not a foreign pattern to me at all. I can even read the error message. It's obvious to me that you're checking whether the brief 'already has a complete process outcome' and erroring out if so. The variable names are descriptive, the check is pretty self explanatory and the error message is right there. That's good code.

Reply:
I had reason to go back and look at the source of my old (much bigger and more complicated) project a few days ago and it was down to comments like this that made me able to comprehend very intricate code that is now quite foreign to me, and the difference was quite stark compared to the opacity and often inexplicability of DM code.

More generally, I suspect it is this tendency to be *this fussy* about all comments that causes our whole source to be so uncommented. When someone hits the hash key they've got to weigh up whether they really want to have to put up with all the bikeshedding that will inevitably come their way.

Reply:
But this comment, specifically, in this specific context, is not concise or helpful.